k2/csrc/fsa_utils.cu

```diff
@@ -2149,4 +2150,407 @@ void FixNumStates(FsaVec *fsas) {
   fsas->shape = RemoveSomeEmptyLists(fsas->shape, 1, renumber_states);
 }
 
+template <typename FloatType>
+Array1<FloatType> GetArcCdf(FsaOrVec &fsas,
+                            Array1<FloatType> &arc_post) {
+  K2_CHECK_GE(fsas.NumAxes(), 2);
+  K2_CHECK_LE(fsas.NumAxes(), 3);
+  int32_t state_axis = (fsas.NumAxes() == 3 ? 1 : 0);
+  ContextPtr c = fsas.Context();
+  int32_t num_states = fsas.TotSize(state_axis),
+      num_arcs = fsas.NumElements();
+  Array1<FloatType> state_post(c, num_states);
+  // Use lowest() instead of -infinity() to initialize the sum, to avoid
+  // -infinity - (-infinity) = NaN; but actually this shouldn't really be an
+  // issue as we shouldn't be taking paths with scores equal to -infinity.
+  LogSumPerSublist(fsas,
+                   std::numeric_limits<FloatType>::lowest(),
+                   &state_post);
+
+  Array1<FloatType> arc_pdf(c, num_arcs);
+  const Arc *arcs_data = fsas.values.Data();
+  const FloatType *arc_post_data = arc_post.Data(),
+      *state_post_data = state_post.Data();
+  FloatType *arc_pdf_data = arc_pdf.Data();
+  // it's row_ids2 if it's an FsaVec (3 axes), else row_ids1.
+  const int32_t *fsas_row_ids2_data = fsas.RowIds(1 + state_axis).Data(),
+      *fsas_row_splits2_data = fsas.RowSplits(1 + state_axis).Data();
+  K2_EVAL(c, num_arcs, lambda_set_arc_probs, (int32_t i) {
+      FloatType arc_post = arc_post_data[i];
+      int32_t state_idx = fsas_row_ids2_data[i];
+      arc_post -= state_post_data[state_idx];
+      arc_pdf_data[i] = exp(arc_post);
+    });
+
+
+  Ragged<FloatType> arc_pdf_ragged(fsas.shape, arc_pdf);
+
+  Array1<FloatType> arc_cdf(c, num_arcs);
+  // TODO: check, the following works with in-place.
+  ExclusiveSumPerSublist(arc_pdf_ragged, 0, &arc_cdf);
```

Comment:
Can `arc_pdf` be reused even if this function supports in-place operation?

The last element of `arc_pdf` will be needed. But if it is modified in place,
then the last element of `arc_pdf` is overwritten and its original value is lost.

Reply:
The last element is 1.0, so we don't need it.

Reply:
Currently, you are using it to compute `inv_tot`.

Reply:
https://github.com/k2-fsa/k2/blob/94049f185e84076ee04a3c844fb8e5926a9fbf30/k2/csrc/fsa_utils.cu#L2215-L2216

Had it been reused, `arc_pdf_data[end_arc - 1]` 's previous value is lost.

Reply:
ah yes you are right, won't make it in place.